deps/spidershim/src/v8string.cc

```diff
@@ -137,6 +137,60 @@ Local<String> String::NewFromTwoByte(Isolate* isolate, const uint16_t* data,
                         length).FromMaybe(Local<String>());
 }
 
+struct ExternalStringFinalizer : JSStringFinalizer {
+  ExternalStringFinalizer(String::ExternalStringResource* resource)
+    : resource_(resource) {}
+  String::ExternalStringResource* resource_;
+
+  // XXX Define finalize() here as a struct member instead of assigning it
+  // after instantiation.
+
+  void dispose() {
+    // Based on V8's Heap::FinalizeExternalString.
+
+    // Dispose of the C++ object if it has not already been disposed.
+    if (this->resource_ != NULL) {
+      this->resource_->Dispose();
+      this->resource_ = nullptr;
+    }
+
+    // Delete ourselves, which JSExternalString::finalize doesn't do,
+    // presumably because it assumes we're static.
+    delete this;
+  };
+};
+
+void FinalizeExternalString(const JSStringFinalizer* fin, char16_t* chars) {
+  const_cast<ExternalStringFinalizer*>(static_cast<const ExternalStringFinalizer*>(fin))->dispose();
+}
+
+MaybeLocal<String> String::NewExternalTwoByte(Isolate* isolate,
+                                              ExternalStringResource* resource) {
+  JSContext* cx = JSContextFromIsolate(isolate);
+
+  auto fin = mozilla::MakeUnique<ExternalStringFinalizer>(resource);
+  if (!fin) {
+    return MaybeLocal<String>();
+  }
+
+  fin->finalize = FinalizeExternalString;
```

Comment:
> It seems like this can be moved to the constructor of ExternalStringFinalizer.

Yes, and that's a better encapsulation of the initialization of the struct. I moved it in c6257b6.